src/main/java/com/aws/greengrass/status/FleetStatusService.java

```diff
@@ -298,38 +306,79 @@ public void schedulePeriodicFleetStatusDataUpdate(boolean isDuringConnectionResu
+            // to wait until all components to reach terminal state to update status
+            uploadFleetStatusServiceData(updatedGreengrassServiceSet, overAllStatus.get(), null,
                     Trigger.COMPONENT_STATUS_CHANGE);
             return;
         }
-
         synchronized (updatedGreengrassServiceSet) {
             updatedGreengrassServiceSet.add(greengrassService);
         }
 
-        // if there is no ongoing deployment and we encounter a BROKEN component, update the fleet status as UNHEALTHY.
-        if (!isDeploymentInProgress.get() && newState.equals(State.BROKEN)) {
-            synchronized (updatedGreengrassServiceSet) {
-                uploadFleetStatusServiceData(updatedGreengrassServiceSet, OverallStatus.UNHEALTHY,
-                        null, Trigger.COMPONENT_STATUS_CHANGE);
+        if (!isDeploymentInProgress.get()) {
+            // if there is no ongoing deployment and we encounter a BROKEN component,
+            // update the fleet status as UNHEALTHY.
+            if (newState.equals(State.BROKEN)) {
+                synchronized (updatedGreengrassServiceSet) {
+                    uploadFleetStatusServiceData(updatedGreengrassServiceSet, OverallStatus.UNHEALTHY,
+                            null, Trigger.COMPONENT_STATUS_CHANGE);
+                }
+            }
+            // Check and update component status when a component reaches to terminal state
+            if (greengrassService.reachedDesiredState() || newState.equals(State.BROKEN)) {
+                handleServiceReachedTerminalState();
             }
         }
     }
 
+    private void handleServiceReachedTerminalState() {
+        //1. Check all other component status
+        List<GreengrassService> servicesToTrack = kernel.orderedDependencies().stream()
+                    .filter(GreengrassService::shouldAutoStart).collect(Collectors.toList());
+        boolean allReachedTerminalState = true;
+        for (GreengrassService service : servicesToTrack) {
+            State state = service.getState();
+            if (!service.reachedDesiredState()) {
+                allReachedTerminalState = false;
+                break;
+            }
+            if (State.RUNNING.equals(state) || State.FINISHED.equals(state) || !service.shouldAutoStart()
+                    && service.reachedDesiredState()) {
+                continue;
+            }
+            if (State.BROKEN.equals(state)) {
+                continue;
+            }
+            allReachedTerminalState = false;
+        }
+        //2. If all other components are in terminal state, update status for all component
+        // Skip updates when kernel shutdown since service ends with FINISHED for stopping
+        if (allReachedTerminalState && !isDeploymentInProgress.get()
+                && !kernelLifecycle.getIsShutdownInitiated().get()) {
+            updateFleetStatusUpdateForAllComponents(Trigger.COMPONENT_STATUS_CHANGE);
```

Comment:
do we update all components? should this be only the components that have changed state?

Reply:
Updated